sdk/core/azure-core/src/main/java/com/azure/core/util/ProgressReporter.java

```diff
@@ -0,0 +1,147 @@
+     */
+    private ProgressReporter(ProgressListener progressListener) {
+        this.progressListener = Objects.requireNonNull(progressListener,
+            "'progressListener' must not be null");
+        this.parent = null;
+    }
+
+    /**
+     * Creates child {@link ProgressReporter}. It tracks it's own progress and reports to parent.
+     * @param parent The parent {@link ProgressReporter}. Must not be null.
+     */
+    private ProgressReporter(ProgressReporter parent) {
+        this.parent = Objects.requireNonNull(parent,
+            "'parent' must not be null");
+        this.progressListener = null;
+    }
+
+    /**
+     * Creates a {@link ProgressReporter} that notifies {@link ProgressListener}.
+     * @param progressListener The {@link ProgressListener} to be notified about progress. Must not be null.
+     * @return The {@link ProgressReporter} instance.
+     * @throws NullPointerException If {@code progressReceiver} is null.
+     */
+    public static ProgressReporter withProgressListener(ProgressListener progressListener) {
+        return new ProgressReporter(progressListener);
+    }
+
+    /**
+     * Creates child {@link ProgressReporter} that can be used to track sub-progress when tracked activity spans
+     * across concurrent processes. Child {@link ProgressReporter} notifies parent about progress and
+     * parent notifies {@link ProgressListener}.
+     * @return The child {@link ProgressReporter}.
+     */
+    public ProgressReporter createChild() {
+        return new ProgressReporter(this);
+    }
+
+    /**
+     * Resets progress to zero and notifies.
+     * <p>
+     * If this is a root {@link ProgressReporter} then attached {@link ProgressListener} is notified.
+     * Otherwise, already accumulated progress is subtracted from the parent {@link ProgressReporter}'s progress.
+     * </p>
+     */
+    public void reset() {
+        long accumulated = PROGRESS_ATOMIC_UPDATER.getAndSet(this, 0L);
+        if (parent != null) {
+            parent.reportProgress(-1L * accumulated);
+        }
+        if (progressListener != null) {
+            progressListener.handleProgress(0L);
+        }
+    }
+
+    /**
+     * Accumulates the provided {@code progress} and notifies.
+     *
+     * <p>
+     * If this is a root {@link ProgressReporter}
+     * then attached {@link ProgressListener} is notified about accumulated progress.
+     * Otherwise, the provided {@code progress} is reported to the parent {@link ProgressReporter}.
+     * </p>
+     *
+     * @param progress The number to be accumulated.
+     */
+    public void reportProgress(long progress) {
+        long totalProgress = PROGRESS_ATOMIC_UPDATER.addAndGet(this, progress);
+        if (parent != null) {
+            parent.reportProgress(progress);
+        }
+        if (progressListener != null) {
+            progressListener.handleProgress(totalProgress);
```

Comment:
The current progress reporting code in java acquires a lock around calling the customer's progress handler code. Is that something that was considered at all here?

Reply:
@rickle-msft Yeah, this particular piece got attention. 
https://github.com/Azure/azure-sdk-for-java/blob/df98d84ad0c7d90d23885d272d40b98a5393ff1b/sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/ProgressReporter.java#L101-L110

The comment there seems to promise that by introducing the lock magically makes user code thread safe and pushes that burden on us... Which is overpromising...

That lock in existing reporters has more traps than benefits, i.e.:
1. with or without lock there's no guarantee of sequencing in paralleled scenarios.
2. we're not assuring any thread safety, customer has to use volatile fields in their handler.
3. if we happen to be on a thread where `reactor.core.scheduler.Schedulers.isInNonBlockingThread()` (or on NIO single threaded event loop) we might run into deadlock/resource starvation.

Therefore we'd rather guide user towards success, by 
1. letting them know that they have to provide thread safe implementation. Here https://github.com/kasobol-msft/azure-sdk-for-java/blob/67837d151e5185897506627de50f5dbefa84abaf/sdk/core/azure-core/src/main/java/com/azure/core/util/ProgressListener.java#L59-L62
2. not locking by default, if user falls into traps above they can mitigate by altering their code.